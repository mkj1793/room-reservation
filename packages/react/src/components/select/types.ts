import {
  ButtonHTMLAttributes,
  DetailedHTMLProps,
  HTMLAttributes,
  LabelHTMLAttributes,
  ReactNode,
  RefObject,
} from 'react';

export type Option = { value: string; label: string; selected?: boolean; isGroupLabel?: boolean; visible?: boolean };
export type Group = { options: Required<Option>[] };
export type SearchResult = Pick<SelectProps, 'groups' | 'options'>;
export type SearchFunction = (searchValue: string, selectedValues: string[], data: SelectData) => Promise<SearchResult>;

export type SelectProps<P = unknown> = {
  options?: (Option | string)[];
  open?: boolean;
  label: string;
  groups?: Array<{
    label: string;
    options: (Option | string)[];
  }>;
  onChange: (selectedValues: string[], data: SelectData) => Partial<SelectData> | void | undefined;
  onSearch?: SearchFunction;
  children?: P | P[];
  multiSelect?: boolean;
  showFiltering?: boolean;
  showSearch?: boolean;
  required?: boolean;
  placeholder?: string;
  id?: string;
  icon?: ReactNode;
  assistiveText?: string;
};

export type SelectData = Required<
  Pick<
    SelectProps,
    'label' | 'open' | 'multiSelect' | 'showFiltering' | 'showSearch' | 'placeholder' | 'required' | 'assistiveText'
  >
> & {
  groups: Array<Group>;
  assistiveText?: string;
  label?: string;
  error?: string;
};

export type SelectMetaData = Pick<SelectProps, 'icon'> & {
  listContainerRef: RefObject<HTMLDivElement>;
  tagListRef: RefObject<HTMLDivElement>;
  showAllButtonRef: RefObject<HTMLButtonElement>;
  selectionButtonRef: RefObject<HTMLButtonElement>;
  searchUpdate: number;
  selectionUpdate: number;
  filter: string;
  search: string;
  isSearching: boolean;
  cancelCurrentSearch: (() => void) | undefined;
  showAllTags: boolean;
  elementIds: {
    button: string;
    label: string;
    container: string;
    tagList: string;
  };
};

export type DivElementProps = DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, never>;
export type LabelElementProps = DetailedHTMLProps<LabelHTMLAttributes<HTMLLabelElement>, never>;
export type ButtonElementProps = DetailedHTMLProps<ButtonHTMLAttributes<HTMLButtonElement>, never>;
export type InputElementProps = DetailedHTMLProps<HTMLAttributes<HTMLInputElement>, never>;
export type UlElementProps = DetailedHTMLProps<HTMLAttributes<HTMLUListElement>, never>;
export type LiElementProps = DetailedHTMLProps<HTMLAttributes<HTMLLIElement>, never>;
export type PropSetterElementTypes = ButtonElementProps | DivElementProps | UlElementProps | LabelElementProps;
