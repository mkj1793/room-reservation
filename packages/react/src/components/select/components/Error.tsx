import React from 'react';

import { Notification } from '../../notification/Notification';
import GroupChild from '../../group/GroupChild';
import { DefaultGroupElementProps } from '../../group/utils';

export function Error(props: DefaultGroupElementProps) {
  return <GroupChild {...props} renderOnlyWithChildren as={(elProps) => <Notification type="error" {...elProps} />} />;
}
