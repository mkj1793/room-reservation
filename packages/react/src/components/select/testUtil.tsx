import { waitFor, fireEvent, render } from '@testing-library/react';
import React from 'react';

import { getLastMockCallArgs } from '../../utils/testHelpers';
import { eventIds, eventTypes } from './events';
import { useSelectDataHandlers } from './hooks/useSelectDataHandlers';
import { Select } from './Select';
import { SelectProps, SelectMetaData, SelectData } from './types';

// storage for the target component to test
const renderMockedComponent = jest.fn();
const renderOnlyTheComponent = jest.fn().mockReturnValue(false);

// tempStorages for data updates triggered from DOM clicks
// These are cleared after each update.
const tempDataStorage = jest.fn();
const tempMetaDataStorage = jest.fn();

export const testUtilBeforeAll = (TargetComponent: React.FC) => {
  renderMockedComponent.mockImplementation(() => <TargetComponent />);
};
export const testUtilAfterAll = () => {
  renderMockedComponent.mockClear();
};

const ButtonsForDataUpdates = () => {
  const dataHandlers = useSelectDataHandlers();

  const onDataClick = () => {
    dataHandlers.updateData(getLastMockCallArgs(tempDataStorage)[0]);
    tempDataStorage.mockReset();
    dataHandlers.trigger({ id: eventIds.showAllButton, type: eventTypes.click });
  };

  const onMetaDataClick = () => {
    dataHandlers.updateMetaData(getLastMockCallArgs(tempMetaDataStorage)[0]);
    tempMetaDataStorage.mockReset();
  };

  return (
    <>
      <button type="button" onClick={onDataClick} id="data-updater">
        Update data
      </button>
      <button type="button" onClick={onMetaDataClick} id="meta-data-updater">
        Update metaData
      </button>
    </>
  );
};

const UpdateChecker = () => {
  // hook usage will cause update when context re-renders
  useSelectDataHandlers();
  return <span id="render-time">{Date.now()}</span>;
};

const ExportedData = () => {
  const { getData } = useSelectDataHandlers();
  const jsonData = JSON.stringify(getData());
  return <span id="exported-data">{jsonData}</span>;
};

const ExportedMetaData = () => {
  const { getMetaData } = useSelectDataHandlers();
  const jsonData = JSON.stringify(getMetaData());
  return <span id="exported-meta-data">{jsonData}</span>;
};

export const mockedContainer = () => {
  if (renderOnlyTheComponent()) {
    return renderMockedComponent();
  }
  return (
    <>
      {renderMockedComponent()}
      <ButtonsForDataUpdates />
      <UpdateChecker />
      <ExportedData />
      <ExportedMetaData />
    </>
  );
};

const onChangeTracker = jest.fn();

const defaultTestProps: SelectProps = {
  options: [],
  onChange: onChangeTracker,
  placeholder: 'Choose one',
  required: false,
  id: 'test-select',
  texts: {
    label: 'Label',
    error: '',
  },
};

export const initTests = ({
  renderComponentOnly,
  selectProps = {},
}: { renderComponentOnly?: boolean; selectProps?: Partial<SelectProps> } = {}) => {
  renderOnlyTheComponent.mockReturnValue(!!renderComponentOnly);
  const props = render(<Select {...{ ...defaultTestProps, ...selectProps }} />);

  const getElementById = <T = HTMLElement,>(id: string) => {
    return props.container.querySelector(`#${id}`) as unknown as T;
  };

  const getRenderTime = () => {
    const el = getElementById('render-time');
    return parseInt(el.innerHTML, 10);
  };

  const createRenderUpdatePromise = () => {
    const lastUpdate = getRenderTime();
    return waitFor(() => {
      if (getRenderTime() <= lastUpdate) {
        throw new Error('Time not updated');
      }
    });
  };

  const getDataFromElement = () => {
    const el = getElementById('exported-data');
    return JSON.parse(el && el.innerHTML);
  };

  const createDataUpdatePromise = () => {
    const lastUpdate = getDataFromElement();
    return waitFor(() => {
      if (getDataFromElement() === lastUpdate) {
        throw new Error('Data not updated');
      }
    });
  };

  const getMetaDataFromElement = () => {
    const el = getElementById('exported-meta-data');
    return JSON.parse(el && el.innerHTML);
  };

  // metadata update does not trigger render!
  const triggerMetaDataChange = async (metaData: Partial<SelectMetaData>) => {
    tempMetaDataStorage(metaData);
    fireEvent.click(getElementById('meta-data-updater'));
  };

  const triggerDataChange = async (data: Partial<SelectData>) => {
    const promise = createRenderUpdatePromise();
    const promise2 = createDataUpdatePromise();
    tempDataStorage(data);
    fireEvent.click(getElementById('data-updater'));
    return Promise.all([promise, promise2]);
  };

  return {
    ...props,
    triggerDataChange,
    triggerMetaDataChange,
    getMetaDataFromElement,
    getDataFromElement,
  };
};
