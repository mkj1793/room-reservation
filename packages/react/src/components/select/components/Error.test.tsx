import { initTests, mockedContainer, testUtilAfterAll, testUtilBeforeAll } from '../testUtil';
import { SelectProps } from '../types';
import { ErrorNotification } from './Error';

jest.mock('./Container', () => {
  return {
    __esModule: true,
    Container: () => mockedContainer(),
  };
});

describe('<ErrorNotification />', () => {
  beforeAll(() => {
    testUtilBeforeAll(ErrorNotification);
  });
  afterAll(() => {
    testUtilAfterAll();
  });
  it('renders the component', () => {
    const { asFragment } = initTests({
      renderComponentOnly: true,
      selectProps: { texts: { error: 'THIS IS AN ERROR' } },
    });
    expect(asFragment()).toMatchSnapshot();
  });
  it('Hello', async () => {
    const selectProps = { texts: { error: 'THIS IS AN ERROR' } };
    const { container, triggerDataChange, triggerMetaDataChange, getMetaDataFromElement } = initTests({ selectProps });
    expect(container).not.toBeNull();
    triggerMetaDataChange({ search: '' });
    selectProps.texts.error = 'NEW ERROR!';
    await triggerDataChange({ open: true });
    expect(getMetaDataFromElement().screenReaderNotifications).toHaveLength(1);
  });
});
