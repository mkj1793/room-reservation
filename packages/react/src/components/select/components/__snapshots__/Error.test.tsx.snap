// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorNotification /> renders the component 1`] = `
<DocumentFragment>
  <div
    class="errorText"
  >
    THIS IS AN ERROR
  </div>
</DocumentFragment>
`;
